tests/e2e/README.md

```diff
@@ -1,4 +1,120 @@
+are already "packaged together" and available as
+`stepsStartChains` in (steps_start_chains.go)[steps_start_chains.go].
+
+An important node is that the parts of the state that are *not* defined are
+just *not checked*. For example, 
+if the balance of a validator isn't listed in a state, it doesn't mean that we want to check
+that the validator balance is absent in the actual system, it means we
+don't care about the balance of that validator in this particular state.
+
+### How to use relayers
+
+For relayers, there are currently several actions defined.
+There are some subtleties here.
+
+* `addChainToRelayer` gives the relayer the config for the chain and should be called before the relayer is expected to interact with the chain for the first time.
+* `createIbcClientsAction`, `addIbcConnectionAction` and `addIbcChannel` are used to create clients, connections and channels.
+* `relayPackets` relays all *currently queued* packets and their acknowledgements.
+* `startRelayer` starts the relayer, which means from now on, it will *automatically relay all packets*.
+Before `startRelayer`, it's possible to let packets queue up and not clear them.
+Afterwards, *all packets* will be relayed immediately, so it's no longer possible
+to e.g. delay the relaying of a packet.
+
+
+## Defining new actions
+
+When we want to test something that does not yet have a corresponding action,
+for example if a feature adds a new transaction to submit on chain and
+there exists no action yet that submits such a tx,
+it's necessary to define a new action.
+You can see the basic template for how to do this by looking at the actions in
+(actions.go)[actions.go].
+The basic principle is to use exec.Command to execute a docker exec to execute a
+command inside the docker container.
+The pattern for this looks something like this:
+```
+cmd := exec.Command("docker", "exec", 
+    tr.containerConfig.InstanceName,
+    tr.chainConfigs[action.Chain].BinaryName,
+    "the command to execute, for example tx",
+    "argument 1",
+    "argument 2")
+output, err := cmd.CombinedOutput()
+if err != nil {
+    log.Fatal(err, "\n", string(bz))
+}
+
+// potentially check something in the output, or log something, ...
+```
+
+Remember that actions don't need to check that the state was modified correctly,
+since we have the state checks for this.
+Still, it's generally a good idea to do a basic check for errors,
+since in case the action encounters an error,
+we'd rather fail inside the action and immediately know something went wrong,
+rather than just getting a non-matching state during the state check that happens after the action.
+
+### Gas
+When submitting txs, generally either specify a large enough amount of gas manually,
+or use `--gas=auto` with a large `--gas-adjustment`.
+You should avoid situations where transactions non-deterministically sometimes
+work and sometimes fail due to gas, as can happen with `--gas=auto` and no `--gas-adjustment`.
```

Comment:
Why is gas consumption non-deterministic? This seems like an important aspect of our e2e testing framework (and probably a reason for our tests sometimes failing). Could you please expand on this section?

Reply:
I don't know why, but you can verify it - run a testnet, then try submitting the same tx again and again with --gas=auto and --gas-adjustment=1. Occasionally, the tx will run out of gas.

I used this script to run a tx again and again:
https://gist.github.com/p-offtermatt/10e128af413fe784f13e1684cd9e8d3e

Here is the relevant output:
```
sh-5.2# ./script.sh 
Running iteration ...
gas estimate: 116693
The string 'out of gas' was found in the output of the second command.
code: 11 codespace: sdk data: "" events: - attributes: - index: true key: fee value: "" - index: true key: fee_payer value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm type: tx - attributes: - index: true key: acc_seq value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm/88 type: tx - attributes: - index: true key: signature value: iejhQ2WHASJr+AgdqITu27ut9i12F/OBsU8piGtR1PseJgCji0eYEAVn15HwFR1gcFUtfFbUp7QTp/n4fW1r4Q== type: tx gas_used: "117243" gas_wanted: "116693" height: "2159" info: "" logs: [] raw_log: 'out of gas in location: WritePerByte; gasWanted: 116693, gasUsed: 117243: out of gas' timestamp: "2023-12-07T08:56:15Z" tx: '@type': /cosmos.tx.v1beta1.Tx auth_info: fee: amount: [] gas_limit: "116693" granter: "" payer: "" signer_infos: - mode_info: single: mode: SIGN_MODE_DIRECT public_key: '@type': /cosmos.crypto.secp256k1.PubKey key: AsFC8tmbGGQSHthsVStbsQ13/+Yza9IT8KCSXXEN7y9f sequence: "88" tip: null body: extension_options: [] memo: "" messages: - '@type': /cosmos.staking.v1beta1.MsgDelegate amount: amount: "500" denom: stake delegator_address: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm validator_address: cosmosvaloper19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddtrgtng non_critical_extension_options: [] timeout_height: "0" signatures: - iejhQ2WHASJr+AgdqITu27ut9i12F/OBsU8piGtR1PseJgCji0eYEAVn15HwFR1gcFUtfFbUp7QTp/n4fW1r4Q== txhash: B8A69A9F51F8458DFFA051974535984356FFE924C5ED3A90B6813E8FE076E860
sh-5.2# ./script.sh 
Running iteration ...
gas estimate: 132915
The string 'out of gas' was not found in the output of the second command.
code: 0 codespace: "" data: 122D0A2B2F636F736D6F732E7374616B696E672E763162657461312E4D736744656C6567617465526573706F6E7365 events: - attributes: - index: true key: fee value: "" - index: true key: fee_payer value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm type: tx - attributes: - index: true key: acc_seq value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm/89 type: tx - attributes: - index: true key: signature value: fqzV7aFZNtrcV489RqFrPyAuTKsHbSNUgKdsLooQjWQPGHjZDw6PwaI1Bh4ubesYA25/WDYmGxnQ9RH/ccOXVA== type: tx - attributes: - index: true key: action value: /cosmos.staking.v1beta1.MsgDelegate - index: true key: sender value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm - index: true key: module value: staking type: message - attributes: - index: true key: spender value: cosmos1jv65s3grqf6v6jl3dp4t6c9t9rk99cd88lyufl - index: true key: amount value: 1383stake type: coin_spent - attributes: - index: true key: receiver value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm - index: true key: amount value: 1383stake type: coin_received - attributes: - index: true key: recipient value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm - index: true key: sender value: cosmos1jv65s3grqf6v6jl3dp4t6c9t9rk99cd88lyufl - index: true key: amount value: 1383stake type: transfer - attributes: - index: true key: sender value: cosmos1jv65s3grqf6v6jl3dp4t6c9t9rk99cd88lyufl type: message - attributes: - index: true key: amount value: 1383stake - index: true key: validator value: cosmosvaloper19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddtrgtng - index: true key: delegator value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm type: withdraw_rewards - attributes: - index: true key: spender value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm - index: true key: amount value: 500stake type: coin_spent - attributes: - index: true key: receiver value: cosmos1fl48vsnmsdzcv85q5d2q4z5ajdha8yu34mf0eh - index: true key: amount value: 500stake type: coin_received - attributes: - index: true key: validator value: cosmosvaloper19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddtrgtng - index: true key: amount value: 500stake - index: true key: new_shares value: "500.000000000000000000" type: delegate gas_used: "130958" gas_wanted: "132915" height: "2161" info: "" logs: - events: - attributes: - key: action value: /cosmos.staking.v1beta1.MsgDelegate - key: sender value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm - key: module value: staking type: message - attributes: - key: spender value: cosmos1jv65s3grqf6v6jl3dp4t6c9t9rk99cd88lyufl - key: amount value: 1383stake type: coin_spent - attributes: - key: receiver value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm - key: amount value: 1383stake type: coin_received - attributes: - key: recipient value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm - key: sender value: cosmos1jv65s3grqf6v6jl3dp4t6c9t9rk99cd88lyufl - key: amount value: 1383stake type: transfer - attributes: - key: sender value: cosmos1jv65s3grqf6v6jl3dp4t6c9t9rk99cd88lyufl type: message - attributes: - key: amount value: 1383stake - key: validator value: cosmosvaloper19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddtrgtng - key: delegator value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm type: withdraw_rewards - attributes: - key: spender value: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm - key: amount value: 500stake type: coin_spent - attributes: - key: receiver value: cosmos1fl48vsnmsdzcv85q5d2q4z5ajdha8yu34mf0eh - key: amount value: 500stake type: coin_received - attributes: - key: validator value: cosmosvaloper19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddtrgtng - key: amount value: 500stake - key: new_shares value: "500.000000000000000000" type: delegate log: "" msg_index: 0 raw_log: '[{"msg_index":0,"events":[{"type":"message","attributes":[{"key":"action","value":"/cosmos.staking.v1beta1.MsgDelegate"},{"key":"sender","value":"cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm"},{"key":"module","value":"staking"}]},{"type":"coin_spent","attributes":[{"key":"spender","value":"cosmos1jv65s3grqf6v6jl3dp4t6c9t9rk99cd88lyufl"},{"key":"amount","value":"1383stake"}]},{"type":"coin_received","attributes":[{"key":"receiver","value":"cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm"},{"key":"amount","value":"1383stake"}]},{"type":"transfer","attributes":[{"key":"recipient","value":"cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm"},{"key":"sender","value":"cosmos1jv65s3grqf6v6jl3dp4t6c9t9rk99cd88lyufl"},{"key":"amount","value":"1383stake"}]},{"type":"message","attributes":[{"key":"sender","value":"cosmos1jv65s3grqf6v6jl3dp4t6c9t9rk99cd88lyufl"}]},{"type":"withdraw_rewards","attributes":[{"key":"amount","value":"1383stake"},{"key":"validator","value":"cosmosvaloper19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddtrgtng"},{"key":"delegator","value":"cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm"}]},{"type":"coin_spent","attributes":[{"key":"spender","value":"cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm"},{"key":"amount","value":"500stake"}]},{"type":"coin_received","attributes":[{"key":"receiver","value":"cosmos1fl48vsnmsdzcv85q5d2q4z5ajdha8yu34mf0eh"},{"key":"amount","value":"500stake"}]},{"type":"delegate","attributes":[{"key":"validator","value":"cosmosvaloper19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddtrgtng"},{"key":"amount","value":"500stake"},{"key":"new_shares","value":"500.000000000000000000"}]}]}]' timestamp: "2023-12-07T08:56:17Z" tx: '@type': /cosmos.tx.v1beta1.Tx auth_info: fee: amount: [] gas_limit: "132915" granter: "" payer: "" signer_infos: - mode_info: single: mode: SIGN_MODE_DIRECT public_key: '@type': /cosmos.crypto.secp256k1.PubKey key: AsFC8tmbGGQSHthsVStbsQ13/+Yza9IT8KCSXXEN7y9f sequence: "89" tip: null body: extension_options: [] memo: "" messages: - '@type': /cosmos.staking.v1beta1.MsgDelegate amount: amount: "500" denom: stake delegator_address: cosmos19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddwhu7lm validator_address: cosmosvaloper19pe9pg5dv9k5fzgzmsrgnw9rl9asf7ddtrgtng non_critical_extension_options: [] timeout_height: "0" signatures: - fqzV7aFZNtrcV489RqFrPyAuTKsHbSNUgKdsLooQjWQPGHjZDw6PwaI1Bh4ubesYA25/WDYmGxnQ9RH/ccOXVA== txhash: 9C779C063F76FAF14964A167D361C311B42449519A5DFCAF4F19456DFE5C556B
```

I expanded the section a bit, please take another look. I'm not sure if there is anything specifically you would like to see here, if so, let me know

Reply:
The section is good. @MSalopek do you have more context on the non-determinism. It's really intriguing